File: typescript/src/amen/resources/customers.ts
import type { AmenClient } from "../client.js";
import type { CreateCustomer, Customer, Page } from "../types.js";
import { toPage } from "./page.js";
export class Customers {
  constructor(private c: AmenClient) {}
  create(body: CreateCustomer) { return this.c.request<Customer>("POST", "/customers/", { json: body }); }
  get(customerNumber: string) { return this.c.request<Customer>("GET", `/customers/${customerNumber}`); }
  async list(params: { page?: number; per_page?: number; type?: string; status?: string } = {}): Promise<Page<Customer>> {
    return toPage<Customer>(await this.c.request("GET", "/customers/", { params }), "customers");
  }
  /** Iterate every page — never process only the first page by accident. */
  async *iterAll(filters: { type?: string; status?: string } = {}): AsyncGenerator<Customer> {
    for (let page = 0; ; page++) {
      const p = await this.list({ ...filters, page });
      yield* p.items;
      if (page + 1 >= p.pages || p.items.length === 0) return;
    }
  }
}

File: flutter/amen_client/lib/src/resources/customers.dart
import '../client.dart';
import '../models.dart';

class Customers {
  final AmenClient _c;
  Customers(this._c);
  Future<Customer> create({required String firstName, required String lastName, required String phoneCode, required String phoneNumber}) async =>
      Customer((await _c.request('POST', '/customers/', json: {'first_name': firstName, 'last_name': lastName, 'phone_code': phoneCode, 'phone_number': phoneNumber}) as Map).cast());
  Future<Customer> get(String customerNumber) async => Customer((await _c.request('GET', '/customers/$customerNumber') as Map).cast());
  Future<Page<Customer>> list({int? page, int? perPage, String? type, String? status}) async =>
      Page.from((await _c.request('GET', '/customers/', params: {'page': page?.toString(), 'per_page': perPage?.toString(), 'type': type, 'status': status}) as Map?)?.cast(), 'customers', Customer.new);
  /// Iterate every page — never process only the first page by accident.
  Stream<Customer> iterAll({String? type, String? status}) async* {
    for (var page = 0;; page++) {
      final p = await list(page: page, type: type, status: status);
      yield* Stream.fromIterable(p.items);
      if (page + 1 >= p.pages || p.items.isEmpty) return;
    }
  }
}
